File: datasets/python_source-code/Multiple-C2/organic_6.py
"""
0/1 Knapsack (DP) - Organic Submission #6
Bottom-up DP that also reconstructs which items were selected.
"""


def knapsack_with_items(weights, values, capacity):
    n = len(weights)
    dp = [[0] * (capacity + 1) for _ in range(n + 1)]

    for i in range(1, n + 1):
        for cap in range(capacity + 1):
            dp[i][cap] = dp[i - 1][cap]
            if weights[i - 1] <= cap:
                candidate = values[i - 1] + dp[i - 1][cap - weights[i - 1]]
                if candidate > dp[i][cap]:
                    dp[i][cap] = candidate

    # backtrack to find which items were taken
    chosen_items = []
    cap = capacity
    for i in range(n, 0, -1):
        if dp[i][cap] != dp[i - 1][cap]:
            chosen_items.append(i - 1)
            cap -= weights[i - 1]

    chosen_items.reverse()
    return dp[n][capacity], chosen_items


if __name__ == "__main__":
    weights = [2, 3, 4, 5]
    values = [3, 4, 5, 6]
    max_value, items = knapsack_with_items(weights, values, 5)
    print(f"Max value: {max_value}, items taken (indices): {items}")

File: server/datasets/controlled_benchmarks/java/multiple_files/multiple_c2/organic_6.java
/*
 * 0/1 Knapsack (DP) - Organic Submission #6
 * Bottom-up DP that also reconstructs which items were selected.
 */

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

class Solution {
    public static int knapsackWithItems(int[] weights, int[] values, int capacity) {
        int n = weights.length;
        int[][] dp = new int[n + 1][capacity + 1];

        for (int i = 1; i <= n; i++) {
            for (int cap = 0; cap <= capacity; cap++) {
                dp[i][cap] = dp[i - 1][cap];
                if (weights[i - 1] <= cap) {
                    int candidate = values[i - 1] + dp[i - 1][cap - weights[i - 1]];
                    if (candidate > dp[i][cap]) {
                        dp[i][cap] = candidate;
                    }
                }
            }
        }

        List<Integer> chosenItems = new ArrayList<>();
        int cap = capacity;
        for (int i = n; i > 0; i--) {
            if (dp[i][cap] != dp[i - 1][cap]) {
                chosenItems.add(i - 1);
                cap -= weights[i - 1];
            }
        }

        Collections.reverse(chosenItems);
        System.out.println("Items taken: " + chosenItems);
        return dp[n][capacity];
    }

    public static void main(String[] args) {
        int[] weights = {2, 3, 4, 5};
        int[] values = {3, 4, 5, 6};
        int maxVal = knapsackWithItems(weights, values, 5);
        System.out.println("Max value: " + maxVal);
    }
}
